File: binding/src/network/libqudpsocket.cpp
#include <QAbstractSocket>
#include <QByteArray>
#include <QHostAddress>
#include <QNetworkDatagram>
#include <QNetworkInterface>
#include <QObject>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QUdpSocket>
#include <qudpsocket.h>
#include "libqudpsocket.h"
#include "libqudpsocket.hxx"

QUdpSocket* QUdpSocket_new() {
	 return new QUdpSocket();
}

QUdpSocket* QUdpSocket_new2(QObject* parent) {
	 return new QUdpSocket(parent);
}

libqt_string QUdpSocket_Tr(const char* s) {
	QString _ret = QUdpSocket::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

bool QUdpSocket_Bind(QUdpSocket* self, int addr) {
	return self->bind(static_cast<QHostAddress::SpecialAddress>(addr));
}

bool QUdpSocket_JoinMulticastGroup(QUdpSocket* self, const QHostAddress* groupAddress) {
	return self->joinMulticastGroup(*groupAddress);
}

bool QUdpSocket_JoinMulticastGroup2(QUdpSocket* self, const QHostAddress* groupAddress, const QNetworkInterface* iface) {
	return self->joinMulticastGroup(*groupAddress, *iface);
}

bool QUdpSocket_LeaveMulticastGroup(QUdpSocket* self, const QHostAddress* groupAddress) {
	return self->leaveMulticastGroup(*groupAddress);
}

bool QUdpSocket_LeaveMulticastGroup2(QUdpSocket* self, const QHostAddress* groupAddress, const QNetworkInterface* iface) {
	return self->leaveMulticastGroup(*groupAddress, *iface);
}

QNetworkInterface* QUdpSocket_MulticastInterface(const QUdpSocket* self) {
	return new QNetworkInterface(self->multicastInterface());
}

void QUdpSocket_SetMulticastInterface(QUdpSocket* self, const QNetworkInterface* iface) {
	self->setMulticastInterface(*iface);
}

bool QUdpSocket_HasPendingDatagrams(const QUdpSocket* self) {
	return self->hasPendingDatagrams();
}

long long QUdpSocket_PendingDatagramSize(const QUdpSocket* self) {
	return self->pendingDatagramSize();
}

QNetworkDatagram* QUdpSocket_ReceiveDatagram(QUdpSocket* self) {
	return new QNetworkDatagram(self->receiveDatagram());
}

long long QUdpSocket_ReadDatagram(QUdpSocket* self, char* data, long long maxlen) {
	return self->readDatagram(data, maxlen);
}

long long QUdpSocket_WriteDatagram(QUdpSocket* self, const QNetworkDatagram* datagram) {
	return self->writeDatagram(*datagram);
}

long long QUdpSocket_WriteDatagram2(QUdpSocket* self, const char* data, long long lenVal, const QHostAddress* host, uint16_t port) {
	return self->writeDatagram(data, lenVal, *host, port);
}

long long QUdpSocket_WriteDatagram3(QUdpSocket* self, const libqt_string datagram, const QHostAddress* host, uint16_t port) {
	return self->writeDatagram(QByteArray(datagram.data, datagram.len), *host, port);
}

libqt_string QUdpSocket_Tr2(const char* s, const char* c) {
	QString _ret = QUdpSocket::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QUdpSocket_Tr3(const char* s, const char* c, int n) {
	QString _ret = QUdpSocket::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

bool QUdpSocket_Bind2(QUdpSocket* self, int addr, uint16_t port) {
	return self->bind(static_cast<QHostAddress::SpecialAddress>(addr), port);
}

bool QUdpSocket_Bind3(QUdpSocket* self, int addr, uint16_t port, int mode) {
	return self->bind(static_cast<QHostAddress::SpecialAddress>(addr), port, static_cast<QFlags<QAbstractSocket::BindFlag>>(mode));
}

QNetworkDatagram* QUdpSocket_ReceiveDatagram1(QUdpSocket* self, long long maxSize) {
	return new QNetworkDatagram(self->receiveDatagram(maxSize));
}

long long QUdpSocket_ReadDatagram3(QUdpSocket* self, char* data, long long maxlen, QHostAddress* host) {
	return self->readDatagram(data, maxlen, host);
}

long long QUdpSocket_ReadDatagram4(QUdpSocket* self, char* data, long long maxlen, QHostAddress* host, uint16_t* port) {
	return self->readDatagram(data, maxlen, host, port);
}

void QUdpSocket_Delete(QUdpSocket* self) {
    delete self;
}

File: binding/src/network/libqudpsocket.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qudpsocket.html
pub const qudpsocket = struct {

    /// New constructs a new QUdpSocket object.
    pub fn New() QtC.QUdpSocket {
        return qtc.QUdpSocket_new();
    }


    /// New2 constructs a new QUdpSocket object.
    pub fn New2(parent: ?*anyopaque) QtC.QUdpSocket {
        return qtc.QUdpSocket_new2(@ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QUdpSocket_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qudpsocket.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Bind(self: ?*anyopaque, addr: i32) bool {
        return qtc.QUdpSocket_Bind(@ptrCast(self), @intCast(addr));
    }

    pub fn JoinMulticastGroup(self: ?*anyopaque, groupAddress: ?*anyopaque) bool {
        return qtc.QUdpSocket_JoinMulticastGroup(@ptrCast(self), @ptrCast(groupAddress));
    }

    pub fn JoinMulticastGroup2(self: ?*anyopaque, groupAddress: ?*anyopaque, iface: ?*anyopaque) bool {
        return qtc.QUdpSocket_JoinMulticastGroup2(@ptrCast(self), @ptrCast(groupAddress), @ptrCast(iface));
    }

    pub fn LeaveMulticastGroup(self: ?*anyopaque, groupAddress: ?*anyopaque) bool {
        return qtc.QUdpSocket_LeaveMulticastGroup(@ptrCast(self), @ptrCast(groupAddress));
    }

    pub fn LeaveMulticastGroup2(self: ?*anyopaque, groupAddress: ?*anyopaque, iface: ?*anyopaque) bool {
        return qtc.QUdpSocket_LeaveMulticastGroup2(@ptrCast(self), @ptrCast(groupAddress), @ptrCast(iface));
    }

    pub fn MulticastInterface(self: ?*anyopaque, ) QtC.QNetworkInterface {
        return qtc.QUdpSocket_MulticastInterface(@ptrCast(self));
    }

    pub fn SetMulticastInterface(self: ?*anyopaque, iface: ?*anyopaque) void {
        qtc.QUdpSocket_SetMulticastInterface(@ptrCast(self), @ptrCast(iface));
    }

    pub fn HasPendingDatagrams(self: ?*anyopaque, ) bool {
        return qtc.QUdpSocket_HasPendingDatagrams(@ptrCast(self));
    }

    pub fn PendingDatagramSize(self: ?*anyopaque, ) i64 {
        return qtc.QUdpSocket_PendingDatagramSize(@ptrCast(self));
    }

    pub fn ReceiveDatagram(self: ?*anyopaque, ) QtC.QNetworkDatagram {
        return qtc.QUdpSocket_ReceiveDatagram(@ptrCast(self));
    }

    pub fn ReadDatagram(self: ?*anyopaque, data: []const u8, maxlen: i64) i64 {
        return qtc.QUdpSocket_ReadDatagram(@ptrCast(self), @ptrCast(data), maxlen);
    }

    pub fn WriteDatagram(self: ?*anyopaque, datagram: ?*anyopaque) i64 {
        return qtc.QUdpSocket_WriteDatagram(@ptrCast(self), @ptrCast(datagram));
    }

    pub fn WriteDatagram2(self: ?*anyopaque, data: []const u8, lenVal: i64, host: ?*anyopaque, port: u16) i64 {
        return qtc.QUdpSocket_WriteDatagram2(@ptrCast(self), @ptrCast(data), lenVal, @ptrCast(host), port);
    }

    pub fn WriteDatagram3(self: ?*anyopaque, datagram: []u8, host: ?*anyopaque, port: u16) i64 {
        const datagram_str = qtc.libqt_string{
    .len = datagram.len,
    .data = datagram.ptr,
};
return qtc.QUdpSocket_WriteDatagram3(@ptrCast(self), datagram_str, @ptrCast(host), port);
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QUdpSocket_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qudpsocket.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QUdpSocket_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qudpsocket.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Bind2(self: ?*anyopaque, addr: i32, port: u16) bool {
        return qtc.QUdpSocket_Bind2(@ptrCast(self), @intCast(addr), port);
    }

    pub fn Bind3(self: ?*anyopaque, addr: i32, port: u16, mode: i32) bool {
        return qtc.QUdpSocket_Bind3(@ptrCast(self), @intCast(addr), port, @intCast(mode));
    }

    pub fn ReceiveDatagram1(self: ?*anyopaque, maxSize: i64) QtC.QNetworkDatagram {
        return qtc.QUdpSocket_ReceiveDatagram1(@ptrCast(self), maxSize);
    }

    pub fn ReadDatagram3(self: ?*anyopaque, data: []const u8, maxlen: i64, host: ?*anyopaque) i64 {
        return qtc.QUdpSocket_ReadDatagram3(@ptrCast(self), @ptrCast(data), maxlen, @ptrCast(host));
    }

    pub fn ReadDatagram4(self: ?*anyopaque, data: []const u8, maxlen: i64, host: ?*anyopaque, port: *u16) i64 {
        return qtc.QUdpSocket_ReadDatagram4(@ptrCast(self), @ptrCast(data), maxlen, @ptrCast(host), @ptrCast(port));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QUdpSocket_Delete(@ptrCast(self));
    }
};
